alphazero/mcts/mcts.py

```diff
@@ -307,7 +326,51 @@ def forward(root_node: MCTSNode):
     # TODO Maybe just iterate through children once, more efficient?
     # Currently doing it twice due to output difference,
     # just do forloop instead?
+    # This can def be optimized better
     return (
-        sum([child.action * child.n_visit for child in start_node.children]),
+        torch.pow(
+            sum(
+                [child.action * child.n_visit for child in start_node.children]
+            ),
+            (1 / temperature),
+        ),
         sampled_node,
     )
+
+
+def self_play(
+    game: GameClassType,
+    net: torch.nn.Module,
+    n_mcts_iter: int = 1600,
+    temperature: float = 1.0,
+    dirichlet_eps: float = 0.25,
+    dirichlet_conc: float = 1.0,
+) -> List[tuple]:
+    pos = game.get_initial_board()
+    data = []
+    is_terminal = game.is_game_over(pos)
+    player = 0
+    node = MCTSNode(state=pos, player=player)
+    n_players = game.n_players()
+    while not is_terminal:
+        policy, node = mcts(
+            start_node=node,
+            game=game,
+            net=net,
+            n_iter=n_mcts_iter,
+            temperature=temperature,
+            dirichlet_eps=dirichlet_eps,
+            dirichlet_conc=dirichlet_conc,
+        )
+        data.append((pos, policy, player))
+        pos = game.board_after_move(
+            pos, player, tuple(node.action.nonzero()[0])
+        )
+        player = (player + 1) % n_players
+        is_terminal = game.is_game_over(pos)
+        # TODO: Currently does not add last state (when game ends, no moves
+        # available) to datapoints,
+        # should it? Also look into stopping thresholds and resigning
```

Comment:
I do not think so, since there is no policy to train on on terminal states.

Reply:
True, I'll remove the comment. Thanks!